File: backend/app/schemas/scan.py
from datetime import datetime
from typing import List, Optional

from pydantic import BaseModel


class SolveRequest(BaseModel):
    subject: Optional[str] = None
    ai_provider: Optional[str] = None
    grade_level: Optional[str] = None


class SolutionStep(BaseModel):
    step: int
    description: str
    formula: Optional[str] = None
    calculation: Optional[str] = None


class FormulaRef(BaseModel):
    id: str
    name: str
    latex: str


class SolutionResponse(BaseModel):
    question_type: str
    knowledge_points: List[str]
    steps: List[SolutionStep]
    final_answer: str
    explanation: Optional[str] = None
    tips: Optional[str] = None


class ScanResponse(BaseModel):
    scan_id: str
    ocr_text: str
    solution: SolutionResponse
    related_formulas: List[FormulaRef]
    created_at: datetime


class ScanRecordResponse(BaseModel):
    id: str
    image_url: Optional[str] = None
    ocr_text: Optional[str]
    subject: Optional[str]
    difficulty: Optional[str]
    created_at: datetime

    class Config:
        from_attributes = True


class FollowUpRequest(BaseModel):
    message: str


class FollowUpResponse(BaseModel):
    reply: str
    message_id: Optional[str] = None
    tokens_used: int = 0


class ConversationMessageResponse(BaseModel):
    id: str
    role: str
    content: str
    created_at: datetime

    class Config:
        from_attributes = True


class ConversationResponse(BaseModel):
    messages: List[ConversationMessageResponse]
    total_messages: int

File: mobile/src/types/scan.ts
export type AIProvider = 'claude' | 'gpt' | 'gemini';

export type Subject =
  | 'math'
  | 'physics'
  | 'chemistry'
  | 'biology'
  | 'english'
  | 'chinese'
  | 'history'
  | 'geography'
  | 'other';

export type Difficulty = 'easy' | 'medium' | 'hard';

export interface SolveRequest {
  image: {
    uri: string;
    type: string;
    name: string;
  };
  subject?: Subject;
  aiProvider?: AIProvider;
  gradeLevel?: string;
}

export interface SolutionStep {
  step: number;
  description: string;
  formula?: string;
  calculation?: string;
}

export interface FormulaRef {
  id: string;
  name: string;
  latex: string;
}

export interface Solution {
  questionType: string;
  knowledgePoints: string[];
  steps: SolutionStep[];
  finalAnswer: string;
  explanation?: string;
  tips?: string;
}

export interface ScanResult {
  scanId: string;
  ocrText: string;
  solution: Solution;
  relatedFormulas: FormulaRef[];
  createdAt: string;
}

export interface ScanRecord {
  id: string;
  imageUrl: string;
  ocrText?: string;
  subject?: Subject;
  difficulty?: Difficulty;
  createdAt: string;
}

export interface ScanState {
  isScanning: boolean;
  isProcessing: boolean;
  currentScan: ScanResult | null;
  error: string | null;
}
